File: leetcode/2352.equal-row-and-column-pairs/solution.py
#!python3
# https://leetcode.com/problems/equal-row-and-column-pairs/submissions/755015576/
from typing import List
from collections import defaultdict

class Solution:
    def equalPairs(self, grid: List[List[int]]) -> int:
        s = defaultdict(lambda: [0, 0])
        n = len(grid)

        for row in grid:
            s[tuple(row)][0] += 1
        for i in range(n):
            col = tuple(row[i] for row in grid)
            s[col][1] += 1

        n = 0
        for _, (r, c) in s.items():
            n += r * c
        return n

File: leetcode/2352.equal-row-and-column-pairs/solution.rs
fn main() {}

struct Solution;
// https://leetcode.com/problems/equal-row-and-column-pairs/
use std::collections::HashMap;

impl Solution {
    pub fn equal_pairs(grid: Vec<Vec<i32>>) -> i32 {
        let n = grid.len();
        let transposed: Vec<Vec<i32>> = (0..n)
            .map(|col| (0..n).map(|row| grid[row][col]).collect())
            .collect();

        let mut row_map: HashMap<Vec<i32>, i32> = HashMap::new();
        for row in grid {
            *row_map.entry(row).or_default() += 1;
        }
        transposed
            .iter()
            .map(|c| row_map.get(c).cloned().unwrap_or_default())
            .sum()
    }
}
